File: Java/populating-next-right-pointers-in-each-node-ii.java
//  Populating Next Right Pointers in Each Node II

/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(1)
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null)
            return;
        TreeLinkNode dummy = new TreeLinkNode(-1);
        for (TreeLinkNode cur = root, pre = dummy; cur != null; cur = cur.next) {
            if (cur.left != null) {
                pre.next = cur.left;
                pre = cur.left;
            }
            if (cur.right != null) {
                pre.next = cur.right;
                pre = cur.right;
            }
        }
        connect(dummy.next);
    }
}

/**
 * Definition for binary tree with next pointer.
 * public class TreeLinkNode {
 *     int val;
 *     TreeLinkNode left, right, next;
 *     TreeLinkNode(int x) { val = x; }
 * }
 */

// Time O(N), Space O(1)
public class Solution {
    public void connect(TreeLinkNode root) {
        if (root == null)
            return;
        TreeLinkNode dummy = new TreeLinkNode(-1);
        dummy.next = root;
        while (dummy.next != null) {
            TreeLinkNode cur_dummy = new TreeLinkNode(-1);
            for (TreeLinkNode cur = dummy.next, pre = cur_dummy; cur != null; cur = cur.next) {
                if (cur.left != null) {
                    pre.next = cur.left;
                    pre = cur.left;
                }
                if (cur.right != null) {
                    pre.next = cur.right;
                    pre = cur.right;
                }
            }
            dummy = cur_dummy;
        }
    }
}

File: leetcode/populating-next-right-pointers-in-each-node-ii.cc
//  Populating Next Right Pointers in Each Node II

/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root)  return;
        TreeLinkNode *dummy = new TreeLinkNode(-1);
        TreeLinkNode *cur = dummy;
        for (TreeLinkNode *p = root; p; p = p->next) {
            if (p->left) {
                cur->next = p->left;
                cur = cur->next;
            }
            if (p->right) {
                cur->next = p->right;
                cur = cur->next;
            }
        }
        connect(dummy->next);
    }
};

class Solution {
public:
    void connect(TreeLinkNode *root) {
        if (!root)  return;
        queue<TreeLinkNode *> q;
        q.push(root);
        while (!q.empty()) {
            int size = q.size();
            for (int i=0; i<size; i++) {
                TreeLinkNode *t = q.front();
                q.pop();
                t->next = (i==size-1) ? nullptr : q.front();
                if (t->left)    q.push(t->left);
                if (t->right)   q.push(t->right);
            }
        }
    }
};
